Feature request: checking files on stdin

Hi there,

I'd like to add support for mypy to Emacs' Flycheck plug-in.  We tend to prefer feeding checkers their input on stdin instead of writing temporary files; is this something that mypy could support too?

The changes should be minimal; probably just an extra flag, similar to `-c` for checking a string; maybe `--stdin`? Additionally, a `--stdin-display-name` flag like flake8 would allow us to tell `mypy` what the name of the original file was.

Thanks!

Hey Guido,

> I'm curious how that would work -- mypy typically has to analyze more than just one file, unless you're using it for a toy program (where it really isn't that useful).

It would work just like `-c`, which is already implemented :)

```
$ echo "x = 5; x = "a"; x = []" > a.py

$ mypy -c "import a; y = 1; y = 'b'"
<string>:1: note: In module imported here:
a.py:1: error: Name 'a' is not defined
a.py:1: error: Incompatible types in assignment (expression has type List[None], variable has type "int")
<string>:1: error: Incompatible types in assignment (expression has type "str", variable has type "int")
```

Flycheck would run mypy from the right directory, so `mypy` would be able to resolve imports.  Additionally, we'd pass `--stdin-display-name` with the file's name to `mypy`, so it really wouldn't be different from reading the actual file.

> I'm also curious how much of a slow-down you'd see if you wrote the data to a temporary file first -- the I/O to do that would seem minimal compared to the cost of starting mypy. Have you tried that yet? And do you have evidence that it's too slow?

There is no slowdown; it's mostly an issue of convenience.  Since `mypy` depends on the file's location, we need to write temporaries to the file's parent folder. Historically, this has caused plenty of issues, including rebuilds from badly configured build systems, leftover temporaries if Emacs doesn't exit cleanly, scripts watching the file system firing too often, and probably others that I don't remember.

I can experiment with a PR, if you want; maybe it'll be easier to decide whether this feature is worth it by looking at actual code?

Thanks for the prompt response!

--- a/docs/source/command_line.rst
+++ b/docs/source/command_line.rst
@@ -419,6 +419,13 @@ Here are some more useful flags:
 - ``--strict`` mode enables all optional error checking flags.  You can see the
   list of flags enabled by strict mode in the full ``mypy -h`` output.
 
+- ``--shadow-file SOURCE_FILE SHADOW_FILE`` makes mypy typecheck SHADOW_FILE in
+  place of SOURCE_FILE.  Primarily intended for tooling.  Allows tooling to
+  make transformations to a file before type checking without having to change
+  the file in-place.  (For example, tooling could use this to display the type
+  of an expression by wrapping it with a call to reveal_type in the shadow
+  file and then parsing the output.)
+
 For the remaining flags you can read the full ``mypy -h`` output.
 
 .. note::

--- a/mypy/main.py
+++ b/mypy/main.py
@@ -323,10 +323,7 @@ def process_options(args: List[str],
                         help=strict_help)
     parser.add_argument('--shadow-file', nargs=2, metavar=('SOURCE_FILE', 'SHADOW_FILE'),
                         dest='shadow_file',
-                        help='Typecheck SHADOW_FILE in place of SOURCE_FILE. '
-                        'Primarily intended for tooling. Allows tooling to make transformations '
-                        'to a file before type checking without having to change the file '
-                        'in-place.')
+                        help='Typecheck SHADOW_FILE in place of SOURCE_FILE.')
     # hidden options
     # --debug-cache will disable any cache-related compressions/optimizations,
     # which will make the cache writing process output pretty-printed JSON (which

--- a/mypy/main.py
+++ b/mypy/main.py
@@ -321,13 +321,13 @@ def process_options(args: List[str],
         ", ".join(strict_flag_names))
     parser.add_argument('--strict', action='store_true', dest='special-opts:strict',
                         help=strict_help)
+    parser.add_argument('--shadow-file', nargs=2, metavar=('SOURCE_FILE', 'SHADOW_FILE'),
+                        dest='shadow_file',
+                        help='Typecheck SHADOW_FILE in place of SOURCE_FILE. '
+                        'Primarily intended for tooling. Allows tooling to make transformations '
+                        'to a file before type checking without having to change the file '
+                        'in-place.')
     # hidden options
-    # --shadow-file a.py tmp.py will typecheck tmp.py in place of a.py.
-    # Useful for tools to make transformations to a file to get more
-    # information from a mypy run without having to change the file in-place
-    # (e.g. by adding a call to reveal_type).
-    parser.add_argument('--shadow-file', metavar='PATH', nargs=2, dest='shadow_file',
-                        help=argparse.SUPPRESS)
     # --debug-cache will disable any cache-related compressions/optimizations,
     # which will make the cache writing process output pretty-printed JSON (which
     # is easier to debug).

